### PR TITLE
Refactor utils/goespatial

### PR DESCRIPTION
**Pull request by sphamba**
# Summary
- put `utils/continents.py` and `utils/countries.py` in `utils/goespatial.py`
- create yaml for continents

### CODE REVIEW

**Comment by sphamba** [comment]
@ghiggi  Do I need to test `remap` and `get_pyresample_area`?

File: gpm_api/tests/test_utils/test_geospatial.py [hunk 1]
```diff
@@ -42,3 +40,378 @@
...
+from pytest_mock import MockFixture
+from typing import Dict, Tuple
+import xarray as xr
+
+from gpm_api.utils import geospatial
+
+ExtentDictionary = Dict[str, Tuple[float, float, float, float]]
+
+
+# @pytest.fixture
```
Comment (ghiggi) [inline gpm_api/tests/test_utils/test_geospatial.py:56]: Remove this if not needed

File: gpm_api/tests/test_utils/test_geospatial.py [hunk 2]
```diff
@@ -42,3 +40,378 @@
...
+
+    # Test with invalid padding
+    with pytest.raises(TypeError):
+        geospatial.get_extent(ds, padding="invalid")
+    with pytest.raises(ValueError):
+        geospatial.get_extent(ds, padding=(0.1,))
+    with pytest.raises(ValueError):
+        geospatial.get_extent(ds, padding=(0.1, 0.2, 0.3))
+
+    # Test with object crossing dateline (not implemented)
```
Comment (ghiggi) [inline gpm_api/tests/test_utils/test_geospatial.py:177]: Yeah this is always a dilemma. The utility is thought to adjust the extent of a cartopy plot. 
I would raise an error suggesting to use: 
- `ccrs.PlateCarree(central_longitude=180)` as cartopy projection 
- and we could potentially provide the `extent` for such cartopy projection (just subtract -180 to the longitudes of the input extent, and recompute the extent)

https://stackoverflow.com/questions/59584276/cartopy-set-extent-with-central-longitude-180

File: gpm_api/tests/test_utils/test_geospatial.py (hunk 1 above)
Comment (ghiggi) [inline gpm_api/tests/test_utils/test_geospatial.py:56]: Remove if not needed

File: gpm_api/tests/test_utils/test_geospatial.py [hunk 3]
```diff
@@ -42,3 +40,378 @@
...
+    # Create dataset
+    ds = xr.Dataset()
+    ds["lat"] = (("cross_track", "along_track"), lat)
+    ds["lon"] = (("cross_track", "along_track"), lon)
+
+    return ds
+
+
+@pytest.fixture
+def grid_dataarray() -> xr.DataArray:
```
Comment (ghiggi) [inline gpm_api/tests/test_utils/test_geospatial.py:224]: We might want to think to move these fixtures in the base `conftest` when developing tests for plotting utility ;)

File: gpm_api/tests/test_utils/test_geospatial.py [hunk 4]
```diff
@@ -42,3 +40,378 @@
...
+    # Create dataset
+    ds = xr.Dataset()
+    ds["lat"] = (("cross_track", "along_track"), lat)
+    ds["lon"] = (("cross_track", "along_track"), lon)
+
+    return ds
+
+
+@pytest.fixture
+def orbit_dataset_multiple_crossings() -> xr.Dataset:
```
Comment (ghiggi) [inline gpm_api/tests/test_utils/test_geospatial.py:205]: `antimeridian_crossing`?

File: gpm_api/tests/test_utils/test_geospatial.py [hunk 5]
```diff
@@ -42,3 +40,378 @@
...
+            geospatial.get_crop_slices_by_extent(da, self.extent)
+
+
+def test_get_crop_slices_by_country(
+    mocker: MockFixture,
+    grid_dataarray: xr.DataArray,
+) -> None:
+    """Test get_crop_slices_by_country"""
+
+    country = "Froopyland"
```
Comment (ghiggi) [inline gpm_api/tests/test_utils/test_geospatial.py:384]: Love this country :)

File: gpm_api/utils/geospatial.py [hunk 6]
```diff
@@ -23,14 +28,114 @@
...
+
+
+def _get_country_extent_dictionary():
+    countries_extent_fpath = os.path.join(_root_path, "gpm_api", "etc", "country_extent.yaml")
+    countries_extent_dict = read_yaml_file(countries_extent_fpath)
+    return countries_extent_dict
+
+
+def get_country_extent(name):
+    # TODO: we could create a dictionary class which
```
Comment (ghiggi) [inline gpm_api/utils/geospatial.py:64]: Remove the TODO ;)

File: gpm_api/utils/geospatial.py [hunk 7]
```diff
@@ -23,14 +28,114 @@
 # - rename file crop.py?
```
Comment (ghiggi) [inline gpm_api/utils/geospatial.py:28]: Do you think it's worth renaming it crop.py? 
I would remove it ...

File: gpm_api/utils/geospatial.py [hunk 8]
```diff
@@ -23,14 +28,114 @@
...
+    # Get country extent dictionary
+    countries_extent_dict = _get_country_extent_dictionary()
+    # Create same dictionary with lower case keys
+    countries_lower_extent_dict = {s.lower(): v for s, v in countries_extent_dict.items()}
+    # Get list of valid countries
+    valid_countries = list(countries_extent_dict.keys())
+    valid_countries_lower = list(countries_lower_extent_dict)
+    if name.lower() in valid_countries_lower:
+        extent = countries_lower_extent_dict[name.lower()]
+        extent = _extend_lonlat_extent(extent, 0.2)
```
Comment (ghiggi) [inline gpm_api/utils/geospatial.py:81]: Maybe we could add this padding as default argument?

File: gpm_api/utils/geospatial.py [hunk 9]
```diff
@@ -23,14 +28,114 @@
...
+
+
+def _get_continent_extent_dictionary():
+    continents_extent_fpath = os.path.join(_root_path, "gpm_api", "etc", "continent_extent.yaml")
+    continents_extent_dict = read_yaml_file(continents_extent_fpath)
+    return continents_extent_dict
+
+
+def get_continent_extent(name):
+    # TODO: we could create a dictionary class which
```
Comment (ghiggi) [inline gpm_api/utils/geospatial.py:99]: Remove this TODO

File: gpm_api/utils/geospatial.py [hunk 10]
```diff
@@ -23,14 +28,114 @@
...
+
+    # Create same dictionary with lower case keys
+    continent_extent_dict = _get_continent_extent_dictionary()
+    continent_lower_extent_dict = {s.lower(): v for s, v in continent_extent_dict.items()}
+    # Get list of valid continents
+    valid_continent = list(continent_extent_dict.keys())
+    valid_continent_lower = list(continent_lower_extent_dict)
+    if name.lower() in valid_continent_lower:
+        extent = continent_lower_extent_dict[name.lower()]
+        # TODO:
```
Comment (ghiggi) [inline gpm_api/utils/geospatial.py:116]: `extent = _extend_lonlat_extent(extent, padding)`
Add `padding=0` as default function argument ...

File: gpm_api/utils/geospatial.py (hunk 8 above)
Comment (ghiggi) [reply gpm_api/utils/geospatial.py:81]: Add `padding=0.2` as default function argument ...

File: gpm_api/utils/geospatial.py [hunk 11]
```diff
@@ -50,7 +155,7 @@ def get_extent(xr_obj, padding=0):
     lat_min = max(-90, np.nanmin(lat).item() - padding[1])
     lat_max = min(90, np.nanmax(lat).item() + padding[1])
 
-    if lon_min > lon_max:
+    if lon_min > lon_max:  # TODO: this condition is never met
```
Comment (ghiggi) [inline gpm_api/utils/geospatial.py:158]: We can have a function `_is_crossing_dateline` checking if two consecutive longitude values are separate by more than 360 degree (-180 --> 180) or viceversa ...  and raise error if this occurs. 
Note that for PMW polar orbit satellites, when the swath overpass the pole we also cross the dateline ...

**Review by ghiggi — APPROVED** [review]
LGTM. 
I just added a couple of comment to clean out residual notes in the code and a suggestion on how to deal with the anti-meridian crossing problem.
Merge it when ready ;)

File: gpm_api/tests/test_utils/test_geospatial.py (hunk 2 above)
Comment (sphamba) [reply gpm_api/tests/test_utils/test_geospatial.py:177]: Perhaps we can add a `central_longitude=0` argument to `get_extent`?

File: gpm_api/tests/test_utils/test_geospatial.py (hunk 3 above)
Comment (sphamba) [reply gpm_api/tests/test_utils/test_geospatial.py:224]: I'll write that down, it depends if these fixture are deemed too specific or not

File: gpm_api/tests/test_utils/test_geospatial.py (hunk 4 above)
Comment (sphamba) [reply gpm_api/tests/test_utils/test_geospatial.py:205]: Multiple passes over of the extent. Should I change the name?

File: gpm_api/utils/geospatial.py (hunk 7 above)
Comment (sphamba) [reply gpm_api/utils/geospatial.py:28]: Well there are also the `get_extent` methods which are exposed, so I would stick with `geospatial` :)

**Comment by sphamba** [comment]
@ghiggi some small questions remaining

File: gpm_api/utils/geospatial.py [hunk 12]
```diff
@@ -135,6 +122,14 @@ def unwrap_longitude_degree(x, period=360):
     return (x + mod) % (2 * mod) - mod
 
 
+def _is_crossing_dateline(lon: Union[list, np.ndarray]):
+    """Check if the longitude array is crossing the dateline."""
+
+    lon = np.asarray(lon)
+    diff = np.diff(lon)
+    return np.any(diff > 180) or np.any(diff < -180)
```
Comment (ghiggi) [inline gpm_api/utils/geospatial.py:130]: ` return np.any(np.abs(diff) > 180)`?

File: gpm_api/utils/geospatial.py [hunk 13]
```diff
@@ -150,15 +145,17 @@ def get_extent(xr_obj, padding: Union[int, float, tuple, list] = 0):
         raise TypeError("Accepted padding type are int, float, list or tuple.")
     lon = xr_obj["lon"].data
     lat = xr_obj["lat"].data
+
+    if _is_crossing_dateline(lon):
+        raise NotImplementedError(
+            "The object cross the dateline. The extent can't be currently be defined."
```
Comment (ghiggi) [inline gpm_api/utils/geospatial.py:151]: currently **be** defined ... remove be 
(I can't suggest code directly ... mystery...)

File: gpm_api/tests/test_utils/test_geospatial.py (hunk 2 above)
Comment (ghiggi) [reply gpm_api/tests/test_utils/test_geospatial.py]: Mmh maybe it's a solution.  
It means we substract the central longitude to the input longitudes?

In the plotting functions, we can try to retrieve the extent. If it raises the `NotImplementedError`, we try with the `get_extent(central_longitude=180)` and to use the the `ccrs.PlateCarree(central_longitude=180)` projection. 

Maybe it's worth to leave it as it is now, and to adapt the function when we define a plotting test case for a swath crossing the dateline and figure out the work-around to plot it nicely ...

File: gpm_api/tests/test_utils/test_geospatial.py (hunk 4 above)
Comment (ghiggi) [reply gpm_api/tests/test_utils/test_geospatial.py:143]: I see that you repeat the array twice. I guess you want to test the dateline crossing.
Maybe you could make the test data more realistic. 
Have latitude ascending and descending (between -90 and 90) and longitude going between -180 and 180 (and repeating). 

You can change the name, i.e. `multiple_antimeridian_crossing_orbit`.

I guess you will end up creating other such sort of orbit fixtures when testing the orbit plotting routines. 
Fine for me to postpone stuffs if you prefer ;)

**Pull request merged**